### ISSUE TITLE
Google error: 405 Method not allowed for url

### ISSUE THREAD

**Issue by IronWolfMC** [pimutils/vdirsyncer#416]
vdirsyncer version: 0.9.4.dev35+ng777eb35
Python: 3.4.3
OS: Ubuntu 14.04 Server (Docker)

Google returns a 405 Client Error (Method Not Allowed for url: https://apidata.googleusercontent.com/caldav/v2/) when attempting to synchronize with a Google calendar. Both the Google Calendar API and the Caldav API are enabled.

Config:
```
[general]
status_path = ~/.config/vdirsyncer/status/

[pair my_calendars]
a = my_cal
b = my_gmail
collections = null

[storage my_cal]
type = caldav
url = http://my.owncloud.server/cloud/remote.php/caldav/calendars/USERNAME/personal
username = USERNAME
password = ****************

[storage my_gmail]
type = google_calendar
token_file = ~/.google-token
client_id = ************************************************.apps.googleusercontent.com
client_secret = ********************
```

Error: (Gist: https://gist.github.com/IronWolfMC/654a22cc583d09322b5d6d183a72371c)
```
root@bbeb76ec35a4:~# vdirsyncer sync
Syncing my_calendars
warning: Skipping '/caldav/v2/', etag property is missing.
Unhandled exception occured for my_calendars.
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/cli/tasks.py", line 66, in sync_collection
    force_delete=force_delete
  File "/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/sync.py", line 216, in sync
    b_info.prepare_idents()
  File "/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/sync.py", line 124, in prepare_idents
    if prefetch else ()):
  File "/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/storage/dav.py", line 423, in get_multi
    headers=self.session.get_default_headers()
  File "/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/storage/dav.py", line 328, in request
    return utils.http.request(method, url, session=self._session, **more)
  File "/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/utils/http.py", line 74, in request
    r.raise_for_status()
  File "/usr/local/lib/python3.4/dist-packages/requests/models.py", line 840, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 405 Client Error: Method Not Allowed for url: https://apidata.googleusercontent.com/caldav/v2/
error: 1 out of 2 tasks failed.
```

**Comment by marmarek** [comment]
Have you enabled caldav access (it is separate from "calendar API") for
that client ID in Google API Manager?

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by IronWolfMC** [comment]
If I've done this correctly, then yes. All of those have been enabled when the error occurred. http://i.imgur.com/RCPCFoO.png

**Comment by untitaker** [comment]
The problem here is that the google storage type doesn't support collection = null, you need to set it to ["personal"] (and also create a calendar with that name on the Google side)

On 8 April 2016 01:52:06 CEST, HunterWolf <notifications@github.com> wrote:
>vdirsyncer version: 0.9.4.dev35+ng777eb35
>Python: 3.4.3
>OS: Ubuntu 14.04 Server (Docker)
>
>Google returns a 405 Client Error (Method Not Allowed for url:
>https://apidata.googleusercontent.com/caldav/v2/) when attempting to
>synchronize with a Google calendar. Both the Google Calendar API and
>the Caldav API are enabled.
>
>Config:
>```
>[general]
>status_path = ~/.config/vdirsyncer/status/
>
>[pair my_calendars]
>a = my_cal
>b = my_gmail
>collections = null
>
>[storage my_cal]
>type = caldav
>url =
>http://my.owncloud.server/cloud/remote.php/caldav/calendars/USERNAME/personal
>username = USERNAME
>password = ****************
>
>[storage my_gmail]
>type = google_calendar
>token_file = ~/.google-token
>client_id =
>************************************************.apps.googleusercontent.com
>client_secret = ********************
>```
>
>Error: (Gist:
>https://gist.github.com/IronWolfMC/654a22cc583d09322b5d6d183a72371c)
>```
>root@bbeb76ec35a4:~# vdirsyncer sync
>Syncing my_calendars
>warning: Skipping '/caldav/v2/', etag property is missing.
>Unhandled exception occured for my_calendars.
>Traceback (most recent call last):
>File
>"/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/cli/tasks.py",
>line 66, in sync_collection
>    force_delete=force_delete
>File
>"/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/sync.py",
>line 216, in sync
>    b_info.prepare_idents()
>File
>"/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/sync.py",
>line 124, in prepare_idents
>    if prefetch else ()):
>File
>"/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/storage/dav.py",
>line 423, in get_multi
>    headers=self.session.get_default_headers()
>File
>"/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/storage/dav.py",
>line 328, in request
>  return utils.http.request(method, url, session=self._session, **more)
>File
>"/usr/local/lib/python3.4/dist-packages/vdirsyncer-0.9.4.dev35+ng777eb35-py3.4.egg/vdirsyncer/utils/http.py",
>line 74, in request
>    r.raise_for_status()
>File "/usr/local/lib/python3.4/dist-packages/requests/models.py", line
>840, in raise_for_status
>    raise HTTPError(http_error_msg, response=self)
>requests.exceptions.HTTPError: 405 Client Error: Method Not Allowed for
>url: https://apidata.googleusercontent.com/caldav/v2/
>error: 1 out of 2 tasks failed.
>```
>
>---
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/issues/416

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by IronWolfMC** [comment]
Thanks, I knew it was just me being dumb, but now I've got a different error message.

After editing the config and creating a Google calendar named personal, I now have this message:

```
root@bbeb76ec35a4:~/.config/vdirsyncer# vdirsyncer sync
Discovering collections for pair my_calendars
error: No collection personal found for storage my_gmail.
Should vdirsyncer attempt to create it? [y/N]: n
critical: Unable to find or create collection "personal" for storage "my_gmail". Please create the collection yourself.
error: 1 out of 1 tasks failed.

root@bbeb76ec35a4:~/.config/vdirsyncer# vdirsyncer sync
Discovering collections for pair my_calendars
error: No collection personal found for storage my_gmail.
Should vdirsyncer attempt to create it? [y/N]: y
error: 400 Client Error: Bad Request for url: https://apidata.googleusercontent.com/caldav/v2/MY_GMAIL_HERE%40gmail.com/personal
critical: Unable to find or create collection "personal" for storage "my_gmail". Please create the collection yourself.
error: 1 out of 1 tasks failed.
```

**Comment by untitaker** [comment]
Yikes.

You're hitting a major wart in vdirsyncer, its extremely limited approach of figuring out which collections of each side should be synced with each other. See #405. While this behavior is usually acceptable with most FOSS DAV servers, it is extremely painful for iCloud, Fastmail, Google (but you can't really blame those either).

The best I can offer you at the moment is to synchronize all Google calendars with a filesystem storage, then have another pair that syncs a particular subfolder with ownCloud.

**Comment by IronWolfMC** [comment]
Ouch. I'll try out the suggestion of syncing with a local storage.

Is this the proper configuration to do that? It still returns the same error as above (no collection found) when it tries to sync from local to Google. Owncloud is syncing to local correctly.

```
[general]
status_path = ~/.config/vdirsyncer/status/

[pair my_calendars]
a = my_cal
b = my_local
collections = ["personal"]

[pair my_google]
a = my_local
b = my_gmail
collections = ["from a"]

[storage my_cal]
type = caldav
url = http://my.owncloud.server/cloud/remote.php/caldav/calendars/USERNAME/personal
username = USERNAME
password = ****************

[storage my_local]
type = filesystem
path = ~/calendar/
fileext = .ics

[storage my_gmail]
type = google_calendar
token_file = ~/.google-token
client_id = ************************************************.apps.googleusercontent.com
client_secret = ********************
```

I have tried both ["from a"] and ["personal"] in the my_google syncing pair. Both return the same error.

**Comment by untitaker** [comment]
Start with this config:


    [general]
    status_path = ~/.config/vdirsyncer/status/

    [pair my_google]
    a = my_local
    b = my_gmail
    collections = ["from b"]

    [storage my_local]
    type = filesystem
    path = ~/calendar/
    fileext = .ics

    [storage my_gmail]
    type = google_calendar
    token_file = ~/.google-token
    client_id = ************************************************.apps.googleusercontent.com
    client_secret = ********************

This will ask to create multiple folders with random-looking strings as folder names. One of them is your personal folder. You'll have to figure out which one is the "personal" one by using grep on the items :(

Append this to the config, replacing FOLDER_NAME_HERE with the folder name of your personal calendar:

    [pair my_calendars]
    a = my_cal
    b = my_local_personal
    collections = null

    [storage my_cal]
    type = caldav
    url = http://my.owncloud.server/cloud/remote.php/caldav/calendars/USERNAME/personal
    username = USERNAME
    password = ****************

    [storage my_local]
    type = filesystem
    path = ~/calendar/FOLDER_NAME_HERE
    fileext = .ics

    [storage my_gmail]
    type = google_calendar
    token_file = ~/.google-token
    client_id = ************************************************.apps.googleusercontent.com
    client_secret = ********************

**Comment by untitaker** [comment]
BTW I'm testing out a UI change right now, so you might want to just hold your breath until I merge something else into master :)

**Comment by IronWolfMC** [comment]
I got it working via creating the initial directory, then adding in my second owncloud calendar which points at a different folder. In folder2, I symlinked the personal calendar to the google calendar's ID in folder1. It synchronizes perfectly.

Thank you so much for your help. I'm excited to see future changes. :)

**Comment by untitaker** [comment]
I'll close this as there's nothing actionable... the larger config issue might be solved together with #405
